### PR TITLE
Contribute to html language server with a custom language.

### LINKED ISSUE #146730: Contribute to html language server with a custom language.
I recreate the old issue from https://github.com/microsoft/vscode/issues/131531 according the comment at https://github.com/microsoft/vscode/issues/49210#issuecomment-1087233035 in order to associate a PR.

//cc @[aeschli](https://github.com/aeschli)

### PR DESCRIPTION
**Pull request by angelozerr**
Contribute to html language server with a custom language.

Signed-off-by: azerr <azerr@redhat.com>

This PR fixes #146730

This PR provides the capability to contribute to the html language server with acustom language (with a simple mean ) with the package.json of the vscode extension.

The `handlebars` custom language is a very good sample and the goal of this PR is to remove the all hard coded of `handlebars` language in the document selector and supported language for auto close.

The handlebars declare in the package.json this contribution:

```json
"html": {
  "languages": [
    {
      "documentSelector": [
        "handlebars"
      ],
      "activateAutoInsertion": true
    }
  ]
}
```

which means handlebars is a language which contribute to the language server with auto insertion activated (auto close tag).

The only thing that I cannot figure out is the force of the start of html language server when handlebars file is opened to remove https://github.com/microsoft/vscode/blob/3698a0b247d7c252b31bd25631647911328ae513/extensions/html-language-features/package.json#L15

In extension which defines an extension.ts I think he can force the start of `vscode.html-language-features` with typecript code.

Any idea are welcome for this problem!

### CODE REVIEW

**Comment by aeschli** [comment]
I would name the contribution point to `html-languages`
(inspired by another contribution point we have called `anycode-languages`)
```
	"contributes": {
		"html-languages": [{
                     "languageId": "xyz-html",
                     "autoInsert": true,
                     ...(more features)
                 }]
         }
```

Right now, such a language will inherit all features of the html-language-feature extension. In the future we can add the capability to select them one by one. Something like:
```
	"highlights": false,
	"completions": false,
	"outline": false,
	"folding": false,
        "formatting": false
	"diagnostics": false
```

Note that the language still needs to provide a grammar for syntax highlighting as well as the language configuration.

On activation. The extension that provides the contribution, also needs to provide `"onLanguage:xyz-html"`.
It needs to have an extension dependency
```	"extensionDependencies": [
		"vscode.html-language-features
	]
```
Because of the language activation point. it needs to provide a `main` and `browser` entry point, which can be empty.


Let me know what you think.

**Comment by angelozerr** [comment]
> I would name the contribution point to html-languages

```json
"contributes": {
		"html-languages": [{
                     "languageId": "xyz-html",
                     "autoInsert": true,
                     ...(more features)
                 }]
         }
```

I like the idea, but today it exists `"html" { "customData": ` My idea was to add a new entry contribution to the existing customData. is it a bad idea? Do you think we should do the same thing than customData and provides this html-languages features in user settings (like customData does)?

> Right now, such a language will inherit all features of the html-language-feature extension. In the future we can add the capability to select them one by one

I love that.

> "extensionDependencies": [
		"vscode.html-language-features
	]

I tried it with handlebars but it didn't work, it's because perhaps there are no extension.ts ? I will retry it.

**Comment by aeschli** [comment]
The norm is to have a single property under `contributes` but no further hierarchy,
The html.customData is the exception and I think I'm going to change it to `htmlCustomData` and `cssCustomData`

I just checked, the majority uses camel case for the property and it is plural.

So for our case: `htmlLanguages`.

**Comment by angelozerr** [comment]
> So for our case: htmlLanguages.

You mean `htmlLanguages` and not `html-languages` ?

**Comment by aeschli** [comment]
Yes, now I think `htmlLanguages` is better than `html-languages`.

**Comment by angelozerr** [comment]
> Yes, now I think htmlLanguages is better than html-languages.

Ok I fixed, handlebars package.json becomes:

```json
"htmlLanguages": [
      {
        "languageId": "handlebars",
        "autoInsert": true
      }
]
```

As handlebars doesn't define a `main` entry with an extension file JS, I cannot declare:

```json
"extensionDependencies": [
  "vscode.html-language-features"
]
```

It means that I need to keep the "onLanguage:handlebars"  in the package.json of vscode.html-language-features. Or perhaps we could define extension.ts which doesn nothing for handlebars?

**Comment by angelozerr** [comment]
Many thanks @aeschli for your great explanation. 

If I understand correctly your comment, I must remove htmlExtendions in handlebars and hard code handlebars like today.

If it that I can update my PR.

**Comment by aeschli** [comment]
There's now new API that allows to access the contributions of extensions across all extension hosts: https://github.com/microsoft/vscode/blob/f3a3cbd829bb62baa71719463e155d7c4919e3ad/src/vscode-dts/vscode.proposed.extensionsAny.d.ts#L37

**Comment by angelozerr** [comment]
> There's now new API that allows to access the contributions of extensions across all extension hosts:

You mean that instead using extensions.all I must use this API ?

> I understand correctly your comment, I must remove htmlExtendions in handlebars and hard code handlebars like today.

Must I update my PR?

**Comment by aeschli** [comment]
So yes, instead of removing htmlExtensions in handlebars you now need to add
```
  "enabledApiProposals": [
    "extensionsAny"
  ],
 ```
to the `package.json` of `html-language-features`
and use  use `extensions.allAcrossExtensionHosts` instead of `extensions.all`

If you are not familiar with that, I can also do that.

**Comment by angelozerr** [comment]
Do you think the PR can be merged ?

> If you are not familiar with that, I can also do that.

Yes it should be really nice!

**Comment by nathan-barrett** [comment]
@angelozerr The PR would be awesome to get merged - anything I can do to help clear any blockers? (i just say a few tests failed but didn't wanna overstep

**Comment by aeschli** [comment]
I made some refactorings
- renamed to language participants
- extracted functionality to LanguageParticipants 
- now using extensions.allAcrossExtensionHosts

I removed the reload dialog for now. Will replace it with a restart of the language client

**Comment by aeschli** [comment]
The contribution point for this feature currently looks like that:
```
    "htmlLanguages": [
      {
        "languageId": "handlebars",
        "autoInsert": true
      }
    ]
```
Alternatives name is `htmlLanguageParticipants`

@jrieken Any objections, preferences?

**Comment by aeschli** [comment]
Changed the contribution name to `htmlLanguageParticipants`

```
    "htmlLanguageParticipants": [
      {
        "languageId": "handlebars",
        "autoInsert": true
      }
    ]
```

**Comment by jrieken** [comment]
Similar cases are `typescriptServerPlugins` and `anycodeLanguages` - I believe there are no strong rules except for camel-case, plural, and scope/topic first. So `htmlLanguageParticipants` seems like a good fit. 

There is the challenge of repeated ids, e.g you can have this 👇 which is can be a bug or a feature. 

```
"htmlLanguageParticipants": [
      {
        "languageId": "handlebars",
        "autoInsert": true
      },
      {
        "languageId": "handlebars",
        "autoInsert": false
      }
]
```

I believe it's OK the way it, other (TS, anycode) have the same issue. Spelling out features is cool, tho `autoInsert` might deserve a better comment - I am not entirely sure what that does.

**Review by jrieken — APPROVED** [review]

File: extensions/html-language-features/schemas/package.schema.json [hunk 1]
```diff
@@ -13,6 +13,23 @@
...
+						"type": "object",
+						"properties": {
+							"languageId": {
+								"type": "string",
+								"description": "The id of the language that participates with HTML language server."
+							},
+							"autoInsert": {
+								"type": "boolean",
+								"description": "Whether the language participates with HTML auto insertions. If not specified, defaults to <code>true</code>."
+							}
```
[the remark this answers is not in the document]
Comment (aeschli) [reply extensions/html-language-features/schemas/package.schema.json:30]: I'd rather have this become a list of all features: 'hover', 'completions', .. and it looks much nicer if they are not negated.

**Comment by angelozerr** [comment]
@aeschli I have intsalled VSCode Insider, the Releases Notes contains this commit, but when I tried to open C:\Users\...\AppData\Local\Programs\Microsoft VS Code Insiders\resources\app\extensions\html-language-features\schemas\package.schema.json it doesn't contains htmlLanguageParticipants definition.

Must we wait some days to have this htmlLanguageParticipants feature in VSCode Insiders? Thanks for your help.

**Comment by aeschli** [comment]
@angelozerr I can see the file in build from today: 13ba7bb446a638d37ebccb1a7d74e31c32bb9790

**Comment by angelozerr** [comment]
@aeschli thanks for your reply. I retried it today and it works perfectly. It is very easy now to provide a HTML support with custom language: https://github.com/redhat-developer/vscode-quarkus/pull/527/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519

Here a little demo with our Qute HTML template support:

QuteHTMLSupport

**Comment by aeschli** [comment]
@angelozerr Cool. Do you want to participate as tester? https://github.com/microsoft/vscode/issues/156200 is our test item and if you want to add a comment on what you have tested then I can add you as tester in the description

**Comment by aeschli** [comment]
@angelozerr I'd add the following section (including your gif) in our release notes.

> The [`vscode-quarkus`](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-quarkus) extension has already adopted the contribution point. It enriches the HTML features with additional proposals for the embedded Qute HTML template language: qute html participant

Is that ok for you?

**Comment by angelozerr** [comment]
It is great @aeschli, thanks so much!

**Pull request merged**